osf/models/preprint.py

```diff
@@ -0,0 +1,920 @@
+from website.project import signals as project_signals
+
+from osf.models.base import BaseModel, GuidMixin
+from osf.models.identifiers import IdentifierMixin, Identifier
+from osf.models.mixins import TaxonomizableMixin, ContributorMixin, SpamOverrideMixin
+from addons.osfstorage.mixins import UploadMixin
+from addons.osfstorage.models import OsfStorageFolder, Region
+
+from framework.sentry import log_exception
+from osf.exceptions import (
+    PreprintStateError, InvalidTagError, TagNotFoundError
+)
+
+logger = logging.getLogger(__name__)
+
+
+class Preprint(DirtyFieldsMixin, GuidMixin, IdentifierMixin, ReviewableMixin, UploadMixin,
+        BaseModel, Loggable, Taggable, GuardianMixin, SpamOverrideMixin, TaxonomizableMixin, ContributorMixin):
+    # Preprint fields that trigger a check to the spam filter on save
+    SPAM_CHECK_FIELDS = {
+        'title',
+        'description',
+    }
+
+    # Setting for ContributorMixin
+    DEFAULT_CONTRIBUTOR_PERMISSIONS = 'write'
+
+    provider = models.ForeignKey('osf.PreprintProvider',
+                                 on_delete=models.SET_NULL,
+                                 related_name='preprints',
+                                 null=True, blank=True, db_index=True)
+    node = models.ForeignKey('osf.AbstractNode', on_delete=models.SET_NULL,
+                             related_name='preprints',
+                             null=True, blank=True, db_index=True)
+    is_published = models.BooleanField(default=False, db_index=True)
+    date_published = NonNaiveDateTimeField(null=True, blank=True)
+    original_publication_date = NonNaiveDateTimeField(null=True, blank=True)
+    license = models.ForeignKey('osf.NodeLicenseRecord',
+                                on_delete=models.SET_NULL, null=True, blank=True)
+
+    identifiers = GenericRelation(Identifier, related_query_name='preprints')
+    preprint_doi_created = NonNaiveDateTimeField(default=None, null=True, blank=True)
+    # begin changes
+    title = models.TextField(
+        validators=[validate_title]
+    )  # this should be a charfield but data from mongo didn't fit in 255
+    description = models.TextField(blank=True, default='')
+    creator = models.ForeignKey(OSFUser,
+                                db_index=True,
+                                related_name='preprints_created',
+                                on_delete=models.SET_NULL,
+                                null=True, blank=True)
+    _contributors = models.ManyToManyField(OSFUser,
+                                           through=PreprintContributor,
+                                           related_name='preprints')
+    article_doi = models.CharField(max_length=128,
+                                            validators=[validate_doi],
+                                            null=True, blank=True)
+    files = GenericRelation('osf.OsfStorageFile', object_id_field='target_object_id', content_type_field='target_content_type')
```

Comment:
Just exposing files relationship, like for nodes. Not an actual field in db. Probably preprints will just have the one primary file, but we are not limiting at the db level how many files/folders can be on the preprint. There can just be one primary file. In the API, we hide any extra files on the preprint.